pkg/util/authorizer/permissiontrie.go

```diff
@@ -0,0 +1,335 @@
+package authorizer
+
+import (
+	"sync"
+)
+
+// Verb constants for bitmask
+const (
+	VerbGet uint32 = 1 << iota
+	VerbList
+	VerbWatch
+	VerbCreate
+	VerbUpdate
+	VerbPatch
+	VerbDelete
+	VerbDeleteCollection
+	VerbImpersonate
+	VerbPost
+	VerbPut
+	VerbAll // Wildcard
+)
+
+// CollectionVerbs represents verbs that act on a collection of resources.
+const CollectionVerbs = VerbList | VerbWatch | VerbCreate | VerbDeleteCollection
+
+// verbMap translates verb strings to their bitmask representation
+var verbMap = map[string]uint32{
+	"get":              VerbGet,
+	"list":             VerbList,
+	"watch":            VerbWatch,
+	"create":           VerbCreate,
+	"update":           VerbUpdate,
+	"patch":            VerbPatch,
+	"delete":           VerbDelete,
+	"deletecollection": VerbDeleteCollection,
+	"impersonate":      VerbImpersonate,
+	"post":             VerbPost,
+	"put":              VerbPut,
+	"*":                VerbAll,
+}
+
+// PermissionTrie holds all permissions in a tree-like structure for fast lookups.
+type PermissionTrie struct {
+	subjectNodes map[string]*SubjectNode
+	mu           sync.RWMutex
+}
```

Comment:
I do not understand how this is a Trie. Can you please explain what happens here? Thanks.